uniffi_bindgen/src/bindings/swift/templates/RustBufferHelper.swift

```diff
@@ -397,16 +418,20 @@ extension Double: Primitive, ViaFfi {
     }
 }
 
-extension Optional: ViaFfiUsingByteBuffer, ViaFfi, Serializable where Wrapped: Serializable {
-    fileprivate static func read(from buf: Reader) throws -> Self {
+{% when Type::Optional  with (inner_type) -%}
+{% let inner_type_name = inner_type|type_swift %}
+
+extension Optional {
+// extension Optional: ViaFfiUsingByteBuffer, ViaFfi, Serializable where Wrapped: Serializable {
```

Comment:
The downside of the approach you're taking here, is that it complicates the expressions for lifting and lowering values of these types. For example, take a look at the `lower_swift` filter, which we use for generating the code that takes a Swift object and lowers it into something to pass across the FFI:

* https://github.com/mozilla/uniffi-rs/blob/d44b96c9de105a6fafc4ff1e1f66e3248631a129/uniffi_bindgen/src/bindings/swift/gen_swift.rs#L226

For almost all values `v` of type `T` this just generates `v.lower()`, so when we want to lower an `Optional<T>` we generate: `v.lower()` and let the Swift compiler figure out how to dispatch it to the right type logic. With the change here that will no longer work, because there is no `lower()` method on `Optional<T>`; you'll instead have to generate `v.lowerUint64` or similar, along the lines of what the case for `Type::Duration` is doing.

Having typed that out...perhaps trying to generate the same generic code as before, but only generate it once, would be a smaller change overall.

Reply:
Yes, this is a very nice explanation of the crossroad at which I found myself. I saw how verbose the generated code would be if I continued down that path and was growing doubtful that my changes would be beneficial. Since then I took the scenic route to opting for adding higher-level methods to `ComponentInterface`. I did so for the `Optional` type and if things look good I'll wrap this up by making the same changes for `Array` and `Dictionary`.

Reply:
Heh, I think the scenic route is very much in the spirit of the meta-goal of this PR so even if we don't like it, it will have been worthwhile :-)

I have a good feeling about it at a high level though, will dig into the details here shortly.